core/src/main/scala/kafka/controller/ControllerChannelManager.scala

```diff
@@ -351,13 +351,17 @@ class ControllerBrokerRequestBatch(controller: KafkaController, stateChangeLogge
     addUpdateMetadataRequestForBrokers(controllerContext.liveOrShuttingDownBrokerIds.toSeq, Set(topicPartition))
   }
 
-  def addStopReplicaRequestForBrokers(brokerIds: Seq[Int], topicPartition: TopicPartition, deletePartition: Boolean,
-                                      callback: (AbstractResponse, Int) => Unit) {
+  def addStopReplicaRequestForBrokers(brokerIds: Seq[Int], topicPartition: TopicPartition, deletePartition: Boolean): Unit = {
     brokerIds.filter(_ >= 0).foreach { brokerId =>
+      def callback(stopReplicaResponse: AbstractResponse): Unit = {
+        if (deletePartition)
+          controller.eventManager.put(controller.StopAndDeleteReplicaResponseReceived(stopReplicaResponse, brokerId))
```

Comment:
This is actually trickier than I thought. In sendRequestsToBrokers(), we only group replicas with no deletion and no callback in a single StopReplicaRequest. By always including a callback, it disables batching in StopReplicaRequest.